app/lib/test_executor.rb

```diff
@@ -0,0 +1,66 @@
+class TestExecutor
+  HOST_WORKDIR = '/usr/src/app'
+  WORKDIR = '/tmp/sandbox'
+
+  def self.images
+    @images ||= {}
+  end
+
+  def self.run_tests(submission)
+    problem = submission.problem
+
+    problem.tests.each do |test|
+      run_test(submission, test)
+    end
+  end
+
+  def self.run_test(submission, test)
+    output = output_of(submission, test)
+
+    if test.expected_output.nil?
+      fill_expected_output(submission.problem, test)
+    end
+
+    output == test.expected_output
```

Comment:
We should probably make a join table to log results to so students can see what happened.